File: scripts/download/download_data.R
library(httr)
library(rvest)
library(jsonlite)

login <- function(username, password) {
  login_url <- "https://auth.biodiversitydata.se/cas/login"
  session <- html_session(login_url)
  form <- html_form(session)[[1]]
  execution_value <- html_nodes(form, xpath = "//input[@name='execution']") %>% html_attr("value")
  payload <- list(
    username = username,
    password = password,
    `_eventId` = 'submit',
    submit = 'LOGIN',
    execution = execution_value
  )
  response <- rvest::submit_form(session, form = form, values = payload)
  if (response$status_code == 200) {
    return(response)
  } else {
    stop("Login failed")
  }
}

get_dataset_list <- function(session) {
  url <- 'https://asv-portal.biodiversitydata.se/list_datasets'
  response <- httr::GET(url)
  if (response$status_code == 200) {
    dataset_list <- jsonlite::fromJSON(content(response, "text"))
    return(dataset_list)
  } else {
    stop("Failed to fetch dataset list")
  }
}

filter_datasets <- function(dataset_list, key, values) {
  if (is.null(values)) {
    return(dataset_list)
  }
  filtered_datasets <- dataset_list[which(dataset_list[[key]] %in% values)]
  return(filtered_datasets)
}

download_datasets <- function(session, datasets) {
  for (dataset in datasets) {
    tryCatch({
      link <- dataset$zip_link
      response <- httr::GET(link)
      if (response$status_code == 200) {
        filename <- paste0(dataset$dataset_id, ".zip")
        writeBin(content(response, "raw"), filename)
        cat("Downloaded:", filename, "\n")
      } else {
        cat("Failed to download", link, "\n")
      }
    }, error = function(e) {
      cat("An error occurred while downloading", link, ":", conditionMessage(e), "\n")
    })
  }
}

main <- function() {
  tryCatch({
    credentials <- jsonlite::read_json('.cas.cred')

    username <- credentials$username
    password <- credentials$password

    session <- login(username, password)

    dataset_list <- get_dataset_list(session)
    cat('\nAVAILABLE DATASETS:\n\n')
    print(dataset_list)

    filtered_list <- filter_datasets(dataset_list, 'target_gene', c('COI'))
    cat('\nSELECTED DATASETS:\n\n')
    print(filtered_list)

    cat('\nSTARTING DOWNLOAD:\n\n')
    download_datasets(session, filtered_list)

  }, error = function(e) {
    cat("An error occurred:", conditionMessage(e), "\n")
  })
}

main()

File: scripts/download/download_data.py
import requests
import json
from bs4 import BeautifulSoup
from typing import Optional, List


def login(username: str, password: str) -> requests.Session:
    """Logs in user and returns a session"""
    login_url = "https://auth.biodiversitydata.se/cas/login"
    soup = BeautifulSoup(requests.get(login_url).content, 'html.parser')
    # First get token
    execution_value = soup.find('input', {'name': 'execution'}).get('value')
    # Then log in
    payload = {
        'username': username,
        'password': password,
        '_eventId': 'submit',
        'submit': 'LOGIN',
        'execution': execution_value
    }
    with requests.Session() as session:
        response = session.post(login_url, data=payload)
        if response.status_code == 200:
            return session
        else:
            raise Exception("Login failed")


def get_dataset_list(session: requests.Session) -> List[dict]:
    """Gets list of datasets available for download from ASV portal"""
    url = 'https://asv-portal.biodiversitydata.se/list_datasets'
    response = session.get(url)
    if response.status_code == 200:
        dataset_list = json.loads(response.text)
        return dataset_list
    else:
        raise Exception("Failed to fetch dataset list")


def filter_datasets(dataset_list: List[dict], key: str,
                    values: Optional[List[str]]) -> List[dict]:
    """Filters datasets on some parameter and value(s), if supplied"""
    if values is None:
        return dataset_list
    filtered_datasets = [ds for ds in dataset_list
                         if key in ds and ds[key] in values]
    return filtered_datasets


def download_datasets(session: requests.Session, datasets: List[dict]):
    """Download dataset zip(s) from ASV portal"""
    for dataset in datasets:
        try:
            link = dataset['zip_link']
            response = session.get(link)
            if response.status_code == 200:
                filename = f"{dataset['dataset_id']}.zip"
                with open(filename, 'wb') as file:
                    file.write(response.content)
                    print(f"Downloaded: {filename}")
            else:
                print(f"Failed to download {link}")
        except Exception as e:
            print(f"An error occurred while downloading {link}: {str(e)}")


def main():
    try:
        with open('.cas.cred') as json_file:
            credentials = json.load(json_file)

        username = credentials["username"]
        password = credentials["password"]

        session = login(username, password)

        dataset_list = get_dataset_list(session)
        print('\nAVAILABLE DATASETS:\n\n')
        print(json.dumps(dataset_list, indent=4))

        filtered_list = filter_datasets(dataset_list, 'target_gene',
                                        ['COI'])
        print('\nSELECTED DATASETS:\n\n')
        print(json.dumps(filtered_list, indent=4))

        print('\nSTARTING DOWNLOAD:\n\n')
        download_datasets(session, filtered_list)

    except Exception as e:
        print(f"An error occurred: {str(e)}")


if __name__ == "__main__":
    main()
